core/modules/wbase/SendChannelShared.cc

```diff
@@ -0,0 +1,223 @@
+/*
+ * LSST Data Management System
+ *
+ * This product includes software developed by the
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+// Class header
+#include "wbase/SendChannelShared.h"
+
+// System headers
+
+// Qserv headers
+#include "global/LogContext.h"
+#include "util/Timer.h"
+
+// LSST headers
+#include "lsst/log/Log.h"
+
+using namespace std;
+
+namespace {
+LOG_LOGGER _log = LOG_GET("lsst.qserv.wbase.SendChannelShared");
+}
+
+namespace lsst {
+namespace qserv {
+namespace wbase {
+
+SendChannelShared::Ptr SendChannelShared::create(SendChannel::Ptr const& sendChannel,
+                                                 wcontrol::TransmitMgr::Ptr const& transmitMgr)  {
+    auto scs = shared_ptr<SendChannelShared>(new SendChannelShared(sendChannel, transmitMgr));
+    return scs;
+}
+
+
+void SendChannelShared::setTaskCount(int taskCount) {
+    _taskCount = taskCount;
+}
+
+
+bool SendChannelShared::transmitTaskLast(StreamGuard sLock, bool inLast) {
+    /// _caller must have locked _streamMutex before calling this.
+    if (not inLast) return false; // This wasn't the last message buffer for this task, so it doesn't matter.
+    ++_lastCount;
+    bool lastTaskDone = _lastCount >= _taskCount;
+    return lastTaskDone;
+}
+
+
+bool SendChannelShared::kill(StreamGuard sLock) {
+    LOGS(_log, LOG_LVL_DEBUG, "SendChannelShared::kill() called");
+    bool ret = _sendChannel->kill();
+    _lastRecvd = true;
+    return ret;
+}
+
+string SendChannelShared::makeIdStr(int qId, int jId) {
+    string str("QID" + (qId == 0 ? "" : to_string(qId) + "#" + to_string(jId)));
+    return str;
+}
+
+
+bool SendChannelShared::addTransmit(bool cancelled, bool erred, bool last, bool largeResult,
+                                    TransmitData::Ptr const& tData, int qId, int jId) {
+    QSERV_LOGCONTEXT_QUERY_JOB(qId, jId);
+    assert(tData != nullptr);
+
+    // This lock may be held for a very long time.
+    std::unique_lock<std::mutex> qLock(_queueMtx);
```

Comment:
I know you said you removed the thread but this makes me think that there can be more threads that call this and they will simply pile up if it goes into a long wait. What makes it wait "long"; waiting for mySQL results?

Reply:
If the czar is being flooded with results, this thread will sit and wait while holding onto the mutex. All the other threads waiting on this mutex would also need to write results back to the same czar, so they'd be stuck too even if there was no mutex. There's really not much that can be done if the czar is busy except wait for it to catch up. 

From past experience, if the workers don't wait for the czar, they crash as they run out of memory holding onto results they can't send.